scripts/cluster-upgrade.sh

```diff
@@ -20,12 +34,140 @@ then
+log() {
+    CD_NAME=$1
+    STAGE=$2
+    MESSAGE=$3
+    
+    echo "$(date "+%Y-%m-%d_%H.%M.%S") - $CD_NAME - $STAGE - $MESSAGE"
+}
+
+setup() {
+    CD_NAMESPACE=$1
+    CD_NAME=$2
+
+    # get kubeconfig so we can check status of cluster's nodes (extra capacity)
+    oc -n $CD_NAMESPACE get secrets "${CD_NAME}-admin-kubeconfig" -o jsonpath='{.data.kubeconfig}' | base64 -d > $TMP_DIR/kubeconfig-$CD_NAME
+
+    ORIGINAL_REPLICAS=$(oc -n $CD_NAMESPACE get clusterdeployment $CD_NAME -o json | jq -r '.metadata.labels["managed.openshift.io/original-worker-replicas"] | select(. != null)')
+    DESIRED_REPLICAS=$(oc -n $CD_NAMESPACE get clusterdeployment $CD_NAME -o json | jq -r '.spec.compute[] | select(.name == "worker") | .replicas')
+    ZONE_COUNT=$(oc -n $CD_NAMESPACE get clusterdeployment $CD_NAME -o json | jq -r '.spec.compute[] | select(.name == "worker") | .platform.aws.zones[]' | wc -l)
+
+    if [ "$ORIGINAL_REPLICAS" == "" ];
+    then
+        # nope, need to bump replicas!
+        ORIGINAL_REPLICAS=$(oc -n $CD_NAMESPACE get clusterdeployment $CD_NAME -o json | jq -r '.spec.compute[0].replicas')
+        DESIRED_REPLICAS=$(($ORIGINAL_REPLICAS+$ZONE_COUNT))
+
+        # update replicas
+        oc -n $CD_NAMESPACE label clusterdeployment $CD_NAME managed.openshift.io/original-worker-replicas=$ORIGINAL_REPLICAS
+        oc -n $CD_NAMESPACE get clusterdeployment $CD_NAME -o json | jq -r ".spec.compute[0].replicas=$DESIRED_REPLICAS" | oc replace -f -
+
+        log $CD_NAME "setup" "bumping replicas from $ORIGINAL_REPLICAS to $DESIRED_REPLICAS"
+    fi
+
+    # make sure we are at capacity in cluster
+    MS_REPLICAS=$(($DESIRED_REPLICAS/$ZONE_COUNT))
+    for MS_NAME in $(KUBECONFIG=$TMP_DIR/kubeconfig-$CD_NAME oc -n openshift-machine-api get machineset --no-headers | grep worker | awk '{print $1}');
+    do
+        log $CD_NAME "setup" "waiting for replicas, machineset=$MS_NAME"
+        AVAILABLE_REPLICAS=$(KUBECONFIG=$TMP_DIR/kubeconfig-$CD_NAME oc -n openshift-machine-api get machineset $MS_NAME -o jsonpath='{.status.availableReplicas}')
+
+        while [ "$AVAILABLE_REPLICAS" != "$MS_REPLICAS" ];
+        do
+            sleep 15
+
+            AVAILABLE_REPLICAS=$(KUBECONFIG=$TMP_DIR/kubeconfig-$CD_NAME oc -n openshift-machine-api get machineset $MS_NAME -o jsonpath='{.status.availableReplicas}')
+        done
+
+        log $CD_NAME "setup" "replicas are ready for upgrade, machineset=$MS_NAME"
+    done
+}
+
+upgrade() {
+    CD_NAMESPACE=$1
+    CD_NAME=$2
+    FROM=$3
+    TO=$4
+    
+    log $CD_NAME "upgrade" "Checking $CD_NAME..."
+
+    # - do we need to upgrade?
+    OCP_CURRENT_VERSION=$(oc -n $CD_NAMESPACE get clusterdeployment $CD_NAME -o json | jq -r '.status.clusterVersionStatus.history[] | select(.state == "Completed") | .version' | head -n1)
+
+    if [ "$OCP_CURRENT_VERSION" == "$TO" ];
+    then
+        log $CD_NAME "upgrade" "skipping, already on version $TO"
+        teardown $CD_NAMESPACE $CD_NAME
```

Comment:
Is this teardown necessary?  It looks like it's only scaling down any extra replicas that were added as part of this process but that doesn't happen until further down in the upgrade function when setup is called, I think?

Reply:
It's there to make sure that teardown is done.  If the label on the clusterdeployment isn't there it will not make any changes.